### PR TITLE
New Cop Style/ClosingArrayBraceLayout

### PR DESCRIPTION
**Pull request by panthomakos**
This cop is inspired by a comment in the `Style/LineBreaks` cop PR:

https://github.com/bbatsov/rubocop/pull/2277#issuecomment-148615911

This cop checks that the closing brace in an array is symmetrical with
respect to the opening brace and the array elements. I am very open to
suggestions on naming and ways to clarify the offense messages.

    # bad
    [ :a,
      :b
    ]

    # bad
    [
      :a,
      :b ]

    # good
    [ :a,
      :b ]

    #good
    [
      :a,
      :b
    ]

My intention is to expand this to account for hashes, method parameters
and method arguments once the language and approach is validated.

### CODE REVIEW

**Comment by bbatsov** [comment]
Technically speaking this is more of a "multi-line array literal layout". A single brace can't have a layout. :-)

**Comment by panthomakos** [comment]
@bbatsov Maybe the name of the cop should be `Style/MultilineArrayBraceLayout` or `Style/MultilineArrayLiteralLayout`? How is the wording of the description?

File: config/disabled.yml [hunk 1]
```diff
@@ -4,6 +4,12 @@ Style/AutoResourceCleanup:
   Description: 'Suggests the usage of an auto resource cleanup version of a method (if available).'
   Enabled: false
 
+Style/ClosingArrayBraceLayout:
+  Description: >-
+                 Checks that the closing brace in an array is symmetrical
```
Comment (bbatsov) [inline config/disabled.yml]: in an array literal

File: lib/rubocop/cop/style/closing_array_brace_layout.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
+# encoding: utf-8
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks that the closing brace in an array is symmetrical
```
Comment (bbatsov) [inline lib/rubocop/cop/style/closing_array_brace_layout.rb]: array literal

File: lib/rubocop/cop/style/closing_array_brace_layout.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+              node.loc.end.begin_pos)
+
+            ->(corrector) { corrector.remove(range) }
+          end
+        end
+
+        private
+
+        def opening_brace_on_same_line?(node)
+          node.loc.begin.line == node.children.first.loc.first_line
```
Comment (bbatsov) [inline lib/rubocop/cop/style/closing_array_brace_layout.rb]: So many chained methods always scare of me of `nil`s. :-)

**Comment by bbatsov** [comment]
I'm fine with both names.

File: lib/rubocop/cop/style/closing_array_brace_layout.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        SAME_LINE_MESSAGE = 'Closing array brace must be on the same line as ' \
+          'the last array element when opening brace is on the same line as ' \
+          'the first array element.'
+
+        NEW_LINE_MESSAGE = 'Closing array brace must be on the line after ' \
+          'the last array element when opening brace is on a separate line ' \
+          'from the first array element.'
+
+        def on_array(node)
+          return unless node.loc.begin
```
Comment (bbatsov) [inline lib/rubocop/cop/style/closing_array_brace_layout.rb]: Normally I add comments on guard clauses, so casual readers would know why it's good idea to exit at this point.

File: lib/rubocop/cop/style/closing_array_brace_layout.rb [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+          else
+            return unless closing_brace_on_same_line?(node)
+
+            add_offense(node, :expression, NEW_LINE_MESSAGE)
+          end
+        end
+
+        def autocorrect(node)
+          if closing_brace_on_same_line?(node)
+            ->(corrector) { corrector.insert_before(node.loc.end, "\n") }
```
Comment (bbatsov) [inline lib/rubocop/cop/style/closing_array_brace_layout.rb]: You might want to freeze the "\n".

**Comment by bbatsov** [comment]
ping

**Comment by panthomakos** [comment]
Holiday/vacation. I promise I'll get back to this.

**Comment by panthomakos** [comment]
@bbatsov I've updated with your comments and renamed to `Style/MultilineArrayBraceLayout`.

**Comment by bbatsov** [comment]
The build is failing due to a style offense.

**Comment by panthomakos** [comment]
@bbatsov fixed

**Pull request merged**